### PR TITLE
Requesting BGZ data from the viewer.

### PR DESCRIPTION
**Pull request by JorisHeadease**
Initially, make this work for Zorgplatform

### CODE REVIEW

File: deployments/dev/hospital/docker-compose.yaml [hunk 1]
```diff
@@ -76,6 +76,7 @@ services:
       - ORCA_CAREPLANCONTRIBUTOR_HEALTHDATAVIEWENDPOINTENABLED=true
       #   Uncomment and set the <...> values below to enable Zorgplatform integration - using zorgplatform production URLs as that is what their developer portal uses
       # - ORCA_CAREPLANCONTRIBUTOR_APPLAUNCH_ZORGPLATFORM_ENABLED=true
+      # - ORCA_CAREPLANCONTRIBUTOR_APPLAUNCH_ZORGPLATFORM_CPSFHIRENDPOINT_URL=http://fhirstore:8080/fhir
```
Comment (JorisHeadease) [inline deployments/dev/hospital/docker-compose.yaml:79]: remove

File: orchestrator/careplancontributor/applaunch/zorgplatform/service.go [hunk 2]
```diff
@@ -165,12 +168,141 @@ type Service struct {
 	zorgplatformCert       *x509.Certificate
 	profile                profile.Provider
 	secureTokenService     SecureTokenService
+	cpsTransport           http.RoundTripper
```
Comment (JorisHeadease) [inline orchestrator/careplancontributor/applaunch/zorgplatform/service.go:171]: can be removed

File: orchestrator/careplancontributor/applaunch/zorgplatform/service.go [hunk 3]
```diff
@@ -165,12 +168,141 @@ type Service struct {
...
+func (s *Service) BgzFhirProxy() coolfhir.HttpProxy {
+	targetFhirBaseUrl, _ := url.Parse(s.config.ApiUrl)
+
+	proxyBasePath := "/cpc/bgz/fhir"
+
+	rewriteUrl, _ := url.Parse(s.baseURL + proxyBasePath)
+
+	return coolfhir.NewProxy("App->EHR (BGZ)", log.Logger, targetFhirBaseUrl, proxyBasePath, rewriteUrl, stsAccessTokenRoundTripper{
+		transport:  s.zorgplatformHttpClient.Transport,
+		zpfService: s,
```
Comment (JorisHeadease) [inline orchestrator/careplancontributor/applaunch/zorgplatform/service.go:187]: can be removed

File: orchestrator/careplancontributor/service.go [hunk 4]
```diff
@@ -137,6 +146,40 @@ func (s Service) RegisterHandlers(mux *http.ServeMux) {
 		}
 		writer.WriteHeader(http.StatusOK)
 	}))
+	// The BgZ aggregate endpoint is used to proxy requests to all CarePlanContributors in the CarePlan. It is used by the HealthDataView to aggregate data from all CarePlanContributors.
+	mux.HandleFunc(fmt.Sprintf("GET %s/aggregate/bgz/fhir/{rest...}", basePath), s.withSessionOrBearerToken(func(writer http.ResponseWriter, request *http.Request) {
```
Comment (JorisHeadease) [inline orchestrator/careplancontributor/service.go:150]: remove bgz from path

File: orchestrator/careplancontributor/service.go [hunk 5]
```diff
@@ -137,6 +146,40 @@ func (s Service) RegisterHandlers(mux *http.ServeMux) {
...
+		if !s.healthdataviewEndpointEnabled {
+			coolfhir.WriteOperationOutcomeFromError(&coolfhir.ErrorWithCode{
+				Message:    "health data view proxy endpoint is disabled",
+				StatusCode: http.StatusMethodNotAllowed,
+			}, fmt.Sprintf("CarePlanContributer/%s %s", request.Method, request.URL.Path), writer)
```
Comment (JorisHeadease) [inline orchestrator/careplancontributor/service.go:155]: reomve

**Review by JorisHeadease — COMMENTED** [review]
added comments from a review session together with Rein. Ended at the CPC service.go around line 300

File: orchestrator/careplancontributor/applaunch/zorgplatform/service.go [hunk 6]
```diff
@@ -165,12 +168,141 @@ type Service struct {
...
+	urlRef.RawQuery = query.Encode()
+
+	// carePlanReference = strings.TrimPrefix(urlRef.Path, "/CarePlan/")
+	startIndex := strings.Index(urlRef.String(), "/CarePlan")
+	if startIndex == -1 {
+		return nil, fmt.Errorf("unable to find CarePlan reference in URL: %s", urlRef.String())
+	}
+
+	localCarePlanRef := urlRef.String()[startIndex:]
+	localCarePlanRef = strings.ReplaceAll(localCarePlanRef, "%3F", "?")
```
Comment (reinkrul) [inline orchestrator/careplancontributor/applaunch/zorgplatform/service.go:226]: looks like instead, the URL should be decoded first, then process

Also, I don't think HTTP header values need escaping? I think regular any ASCII character (except newlines) are allowed.

File: orchestrator/careplancontributor/applaunch/zorgplatform/service.go [hunk 7]
```diff
@@ -165,12 +168,141 @@ type Service struct {
...
+	var bsnIdentifier *fhir.Identifier
+	for _, identifier := range patient.Identifier {
+		if identifier.System != nil && *identifier.System == "http://fhir.nl/fhir/NamingSystem/bsn" {
+			bsnIdentifier = &identifier
+			break
+		}
+	}
+	if bsnIdentifier == nil {
+		return nil, fmt.Errorf("identifier with system http://fhir.nl/fhir/NamingSystem/bsn not found")
+	}
```
Comment (reinkrul) [inline orchestrator/careplancontributor/applaunch/zorgplatform/service.go:278]: better use/introduce a FilterIdentifier in coolfhir

File: orchestrator/careplancontributor/service.go [hunk 8]
```diff
@@ -317,7 +457,7 @@ func (s Service) withSessionOrBearerToken(next func(response http.ResponseWriter
 			next(response, request)
 			return
 		}
-		http.Error(response, "no session found", http.StatusUnauthorized)
+		http.Error(response, "no session or bearer token found", http.StatusUnauthorized)
```
Comment (reinkrul) [inline orchestrator/careplancontributor/service.go:460]: static bearer token is probably not something we want to keep, so I'd like to keep the original error message.

**Pull request merged**